src/snabbdom.ts

```diff
@@ -1,55 +1,88 @@
-/* global module, document, Node */
-import {Module} from './modules/module';
-import {Hooks} from './hooks';
-import vnode, {VNode, VNodeData, Key} from './vnode';
-import * as is from './is';
-import htmlDomApi, {DOMAPI} from './htmldomapi';
-
-function isUndef(s: any): boolean { return s === undefined; }
+/* global Node */
+import {VNodeType, Hooks, InsertHook, VNode, VNodes, Key, VText, VComment,
+        Module, VElement, VFragment, DOMAPI, PreHook, CreateHook, UpdateHook, DestroyHook, RemoveHook, PostHook} from './types';
+
+const { isArray } = Array;
+
+function isUndef(s: any): s is undefined { return s === undefined; }
 function isDef(s: any): boolean { return s !== undefined; }
 
-type VNodeQueue = Array<VNode>;
+type VNodeQueue = VNode[];
+
+const emptyNode: VNode = {
+  nt: 0,
+  sel: '',
+  data: {},
+  children: undefined,
+  text: undefined,
+  elm: undefined,
+  key: undefined,
+};
 
-const emptyNode = vnode('', {}, [], undefined, undefined);
+type KeyToIndexMap = {[key: string]: number};
+
+type ArraysOf<T> = {
+  [K in keyof T]: (T[K])[];
+}
+
+export interface PatchFunction {
+  (oldVnode: VNode, vnode: VNode): VNode;
+  children(parentElm: Node, oldCh: VNodes, newCh: VNodes): VNodes;
+}
 
 function sameVnode(vnode1: VNode, vnode2: VNode): boolean {
-  return vnode1.key === vnode2.key && vnode1.sel === vnode2.sel;
+  return vnode1.nt === vnode2.nt && vnode1.key === vnode2.key && vnode1.sel === vnode2.sel;
 }
 
-function isVnode(vnode: any): vnode is VNode {
-  return vnode.sel !== undefined;
+function isVNode(vnode: any): vnode is VNode {
+  return vnode != null;
 }
 
-type KeyToIndexMap = {[key: string]: number};
+function isElementVNode(vnode: VNode): vnode is VElement {
+  return vnode.nt === VNodeType.Element;
+}
```

Comment:
This provoke a regression when tested against current code base because my initial vnode dont have `nt` property. Always `undefined`. When this should be set ?

Reply:
Yes, this is a breaking change if you are building your own vnodes (you will need to change the shape of those vnodes to include this new property), but if you use our `h()` or `toVNode()`, we will take care of it.

Reply:
I'm using snabbdom-jsx so that's why :)

Reply:
we will need to have a new version for that one, once this is ready. feel free to branch out snabbdom-jsx, to adjust the produced shaped.